File: src/data_rover/api/routes/rules.py
"""Rules lint: parse + schema + drift check for the editor's debounced calls.

Sibling of POST /metamodel/lint: cheap (no model iteration, no write_mutex),
deliberately NOT in the read-only-POST allowlist — only the editing flow
lints, and viewers have nothing to lint.

Always-200 covers LINT RESULTS, not the request envelope: any well-formed
body (a string ``yaml``, of any content — unparseable YAML, schema-violating
YAML, drifted rules, a bare scalar, a list, an empty document) reaches
``lint_rules`` and gets a 200. A malformed envelope (missing/non-string
``yaml``, invalid JSON, an oversized ``yaml``) never reaches it — FastAPI's
request validation 422s first, because the envelope is a client contract,
not a lint candidate, unlike ``/metamodel/lint`` whose raw body IS the
candidate.

``POST /rules/parse`` answers the same way for the engine: the rule set's
normalized document, or the one parse error lint gives. It reads no session
and no metamodel, so it never hydrates the project; like lint, a viewer gets
a 403."""

from __future__ import annotations

import json

from fastapi import APIRouter, Depends

from data_rover.core.validation.rules.compile import RuleSetSource, compile_rule_sets
from data_rover.core.validation.rules.schema import (
    RuleSetDefinition,
    RuleSetError,
    parse_rule_set,
)

from ..authz import require_membership
from ..db_models import Membership
from ..deps import Session, get_request_session
from ..schemas import (
    LintErrorOut,
    RulesLintRequest,
    RulesLintResponse,
    RulesParseOut,
    RulesParseRequest,
    RuleWarningOut,
)

router = APIRouter()


def _lint_error(exc: RuleSetError) -> LintErrorOut:
    mark = getattr(exc.__cause__, "problem_mark", None)
    return LintErrorOut(
        message=str(exc),
        line=mark.line + 1 if mark is not None else None,
        column=mark.column + 1 if mark is not None else None,
    )


def rules_document(defn: RuleSetDefinition) -> str:
    """The rule set as the engine reads it: aliased keys, only the fields the
    author wrote, in field order. ``Infinity`` and ``NaN`` stay bare
    literals, floats keep their ``.0`` and integers stay exact."""
    return json.dumps(
        defn.model_dump(mode="python", by_alias=True, exclude_unset=True),
        ensure_ascii=False,
        separators=(",", ":"),
    )


def parse_result(yaml: str) -> RulesParseOut:
    try:
        defn = parse_rule_set(yaml)
    except RuleSetError as exc:
        return RulesParseOut(ok=False, errors=[_lint_error(exc)])
    return RulesParseOut(ok=True, document=rules_document(defn))


@router.post("/rules/lint")
def lint_rules(
    payload: RulesLintRequest,
    session: Session = Depends(get_request_session),
) -> RulesLintResponse:
    try:
        parse_rule_set(payload.yaml)
    except RuleSetError as exc:
        return RulesLintResponse(ok=False, errors=[_lint_error(exc)])
    warnings: list[RuleWarningOut] = []
    if session.metamodel is not None:
        compiled = compile_rule_sets(
            [RuleSetSource("draft", "draft", payload.yaml)], session.metamodel
        )
        warnings = [
            RuleWarningOut(rule=d.rule, message=d.reason) for d in compiled.skipped
        ]
    return RulesLintResponse(ok=True, warnings=warnings)


@router.post("/rules/parse")
def parse_rules(
    payload: RulesParseRequest,
    _membership: Membership = Depends(require_membership),
) -> RulesParseOut:
    return parse_result(payload.yaml)

File: frontend/src/lib/api/rules.ts
import { z } from 'zod';
import { apiFetch, type ClientConfig } from './client';
import { MetamodelLintErrorSchema } from './types';

/** One rules-lint finding. Position is best-effort and 1-based: only a YAML
 * PARSE error carries a line/column (from the parser mark); a schema violation
 * is message-only.
 *
 * The metamodel linter answers the same shape off the same server-side model,
 * so the two share ONE schema rather than two copies that can drift. */
export const RulesLintErrorSchema = MetamodelLintErrorSchema;
export type RulesLintError = z.infer<typeof RulesLintErrorSchema>;

/** Drift: a rule naming a stereotype, relationship or property the metamodel
 * does not have. A WARNING, never an error — the metamodel can change under a
 * committed rule set, so a drifted rule is skipped, not invalid. */
export const RulesLintWarningSchema = z.object({
	rule: z.string(),
	message: z.string()
});
export type RulesLintWarning = z.infer<typeof RulesLintWarningSchema>;

export const RulesLintSchema = z.object({
	ok: z.boolean(),
	errors: z.array(RulesLintErrorSchema).default([]),
	warnings: z.array(RulesLintWarningSchema).default([])
});
export type RulesLint = z.infer<typeof RulesLintSchema>;

/**
 * Parse + schema + drift check for the rules editor's debounced calls.
 *
 * Sibling of `lintMetamodel`, with one difference that matters to callers: the
 * YAML rides in a JSON ENVELOPE rather than being the raw body, so any lint
 * RESULT (unparseable, schema-violating, drifted) comes back 200 while a
 * rejected envelope is a 422 — in practice a document past the server's size
 * cap, since the envelope itself is always well-formed here.
 */
export function lintRules(yaml: string, cfg?: ClientConfig): Promise<RulesLint> {
	return apiFetch('/rules/lint', { method: 'POST', body: { yaml }, schema: RulesLintSchema }, cfg);
}
